### PR TITLE
os/bluestore: fix remove_collection to properly detect collection e…

### PR DESCRIPTION
**Pull request by ifed01**
…mptiness when onodes aren't in memory

Signed-off-by: Igor Fedotov <ifedotov@mirantis.com>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -8653,25 +8665,54 @@ int BlueStore::_remove_collection(TransContext *txc, coll_t cid,
...
+
+    vector<ghobject_t> ls;
+    ghobject_t next;
+    //Enumerate onodes in db, up to nonexistent_count + 1
+    // then check if all of them are marked as non-existent.
+    // Bypass the check if returned number is greater than nonexistent_count
+    r = _collection_list(c->get(), ghobject_t(), ghobject_t::get_max(), true, nonexistent_count + 1,
+                          &ls, &next);
+
+    if ( r>=0 ) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: if (r >= 0) {

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -8653,25 +8665,54 @@ int BlueStore::_remove_collection(TransContext *txc, coll_t cid,
...
+      bool exists = false; //ls.size() > nonexistent_count;
+      for( auto it = ls.begin(); !exists && it < ls.end(); ++it) {
+       dout(10)<<__func__<<" oid " << *it << dendl;
+        auto onode = (*c)->onode_map.lookup(*it);
+        exists = !onode || onode->exists;
+        if (exists) {
+          dout(10) << __func__ << " " << *it
+                   << " exists in db" << dendl;
+	}
+      }
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: indentation here is all weird?

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -8653,25 +8665,54 @@ int BlueStore::_remove_collection(TransContext *txc, coll_t cid,
...
+    //Enumerate onodes in db, up to nonexistent_count + 1
+    // then check if all of them are marked as non-existent.
+    // Bypass the check if returned number is greater than nonexistent_count
+    r = _collection_list(c->get(), ghobject_t(), ghobject_t::get_max(), true, nonexistent_count + 1,
+                          &ls, &next);
+
+    if ( r>=0 ) {
+      bool exists = false; //ls.size() > nonexistent_count;
+      for( auto it = ls.begin(); !exists && it < ls.end(); ++it) {
+       dout(10)<<__func__<<" oid " << *it << dendl;
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: missing whitespace

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -8653,25 +8665,54 @@ int BlueStore::_remove_collection(TransContext *txc, coll_t cid,
...
+    ghobject_t next;
+    //Enumerate onodes in db, up to nonexistent_count + 1
+    // then check if all of them are marked as non-existent.
+    // Bypass the check if returned number is greater than nonexistent_count
+    r = _collection_list(c->get(), ghobject_t(), ghobject_t::get_max(), true, nonexistent_count + 1,
+                          &ls, &next);
+
+    if ( r>=0 ) {
+      bool exists = false; //ls.size() > nonexistent_count;
+      for( auto it = ls.begin(); !exists && it < ls.end(); ++it) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: misplaced whitesapce

**Comment by ifed01** [comment]
@liewegas fixed.

**Pull request merged**